src/niMidi.cpp

```diff
@@ -191,28 +295,87 @@ class NiMidiSurface: public BaseSurface {
 				this->_onKnobPanChange(command, convertSignedMidiValue(value));
 				break;
 			default:
+#ifdef BASIC_DIAGNOSTICS
 				ostringstream s;
 				s << "Unhandled MIDI message " << showbase << hex
 					<< (int)event->midi_message[0] << " "
 					<< (int)event->midi_message[1] << " "
 					<< (int)event->midi_message[2] << endl;
 				ShowConsoleMsg(s.str().c_str());
+#endif
 				break;
 		}
 	}
 
+	void _peakMixerUpdate() override {
+		// Peak meters. Note: Reaper reports peak, NOT VU	
+
+		// ToDo: Peak Hold in KK display shall be erased immediately when changing bank
+		// ToDo: Peak Hold in KK display shall be erased after decay time t when track muted or no signal.
+		// ToDo: Explore the effect of sending CMD_SEL_TRACK_PARAMS_CHANGED after sending CMD_TRACK_VU
+		
+		// Meter information is sent to KK as array (string of chars) for all 16 channels (8 x stereo) of one bank.
+		// A value of 0 will result in stopping to refresh meters further to right.
+		// The array needs one additional char at peakBank[16] set to 0 as a "end of string" marker.
+		char peakBank[(BANK_NUM_TRACKS * 2) + 1] = { 0 };
+		int j = 0;
+		double peakValue = 0;		
+		int numInBank = 0;
+		int bankEnd = this->_bankStart + BANK_NUM_TRACKS - 1;
+		int numTracks = CSurf_NumTracks(false); // If we ever want to show just MCP tracks in KK Mixer View (param) must be (true)
+		if (bankEnd > numTracks) {
+			bankEnd = numTracks;
+		}
+		for (int id = this->_bankStart; id <= bankEnd; ++id, ++numInBank) {
+			MediaTrack* track = CSurf_TrackFromID(id, false);
+			if (!track) {
+				break;
+			}
+			j = 2 * numInBank;
+			// Muted tracks still report peak levels => ignore these
+			if (*(bool*)GetSetMediaTrackInfo(track, "B_MUTE", nullptr)) {
+				peakBank[j] = 1;
+				peakBank[j+1] = 1;
+			}
+			else {
+				peakValue = Track_GetPeakInfo(track, 0); // left channel
+				if (peakValue < 0.0000000298023223876953125) {
```

Comment:
1. Please use a named constant for this value. I assume it's the float representation of -150dB?
2. I wonder if it makes sense to move this check into peakToChar?

Reply:
yup. I did exactly that in later rework of the code (see later pull requests). And yes, sorry for making you read through these ventures with the incremental modifications style....